File: LR3_Ilichev/Sportsman.cpp
#include "Sportsman.h"

BOOST_CLASS_VERSION(Sportsman, 1)

using namespace std;

int Sportsman::max_id = 1;

void Sportsman::reset_maxID()
{
    max_id = 1;
}

std::shared_ptr<Sportsman> Sportsman::clone() const{
    return std::make_shared<Sportsman>(*this);
}

void Sportsman::update(QStringList& data){
    surname = data[0].toLocal8Bit();
    name = data[1].toLocal8Bit();
    age = data[2].toInt();
    height = data[3].toInt();
}

QStringList Sportsman::get_data() const
{
    return QStringList({QString::fromLocal8Bit(surname.c_str()),
                        QString::fromLocal8Bit(name.c_str()),
                        QString::number(age),
                        QString::number(height)});
}

void Sportsman::draw(QPainter *painter, int x, int& y, QVector<int>& column_widths, int padding, int height) const {

    QStringList data = get_data();
    draw_table_row(painter, column_widths, data, x, y, padding, height);
    y += height;
}

QVector<int> Sportsman::get_column_widths(QPainter* painter) const
{
    QStringList data = get_data();
    QVector<int> widths(data.size(), 0);

    for (int i = 0; i < data.size(); i++)
        widths[i] = painter->fontMetrics().horizontalAdvance(data[i]);

    return widths;
}

File: С#_Desktop_Application/C#_Form/MyAppIlichev/Sportsman.cs
using System.Runtime.InteropServices;


namespace MyAppIlichev
{
    internal class Sportsman
    {
        private static int MaxId;
        public int Id { get; set; }
        public string Surname { get; set; }
        public string Name { get; set; }
        public int Age { get; set; }
        public int Height { get; set; }

        public Sportsman()
        {
            Id = ++MaxId;
            Name = "";
            Surname = "";
            Age = 0;
            Height = 0;
        }

        public virtual SportsmanDataStruct GetStruct()
        {
            SportsmanDataStruct SD = new SportsmanDataStruct
            {
                id = Id,
                surname = Surname,
                name = Name,
                age = Age,
                height = Height
            };

            return SD;
        }

        public virtual List<string> GetData()
        {
            List<string> data = new List<string>
            {
                Surname,
                Name,
                Age.ToString(),
                Height.ToString()
            };

            return data;
        }


        public static void ResetMaxId() { MaxId = 0; }
    }
}
